sample-resource-plugin/src/main/java/org/opensearch/sample/resource/actions/transport/GetResourceTransportAction.java

```diff
@@ -0,0 +1,173 @@
+import org.opensearch.common.xcontent.XContentType;
+import org.opensearch.core.action.ActionListener;
+import org.opensearch.core.common.Strings;
+import org.opensearch.core.rest.RestStatus;
+import org.opensearch.core.xcontent.NamedXContentRegistry;
+import org.opensearch.core.xcontent.XContentParser;
+import org.opensearch.index.query.QueryBuilders;
+import org.opensearch.sample.SampleResource;
+import org.opensearch.sample.resource.actions.rest.get.GetResourceAction;
+import org.opensearch.sample.resource.actions.rest.get.GetResourceRequest;
+import org.opensearch.sample.resource.actions.rest.get.GetResourceResponse;
+import org.opensearch.sample.resource.client.ResourceSharingClientAccessor;
+import org.opensearch.search.SearchHit;
+import org.opensearch.search.builder.SearchSourceBuilder;
+import org.opensearch.security.spi.resources.client.ResourceSharingClient;
+import org.opensearch.tasks.Task;
+import org.opensearch.transport.TransportService;
+import org.opensearch.transport.client.node.NodeClient;
+
+import static org.opensearch.sample.utils.Constants.RESOURCE_INDEX_NAME;
+
+/**
+ * Transport action for getting a resource
+ */
+public class GetResourceTransportAction extends HandledTransportAction<GetResourceRequest, GetResourceResponse> {
+    private static final Logger log = LogManager.getLogger(GetResourceTransportAction.class);
+
+    private final TransportService transportService;
+    private final NodeClient nodeClient;
+    private final Settings settings;
+
+    @Inject
+    public GetResourceTransportAction(
+        Settings settings,
+        TransportService transportService,
+        ActionFilters actionFilters,
+        NodeClient nodeClient
+    ) {
+        super(GetResourceAction.NAME, transportService, actionFilters, GetResourceRequest::new);
+        this.transportService = transportService;
+        this.nodeClient = nodeClient;
+        this.settings = settings;
+    }
+
+    @Override
+    protected void doExecute(Task task, GetResourceRequest request, ActionListener<GetResourceResponse> listener) {
+        ResourceSharingClient client = ResourceSharingClientAccessor.getInstance().getResourceSharingClient();
+        String resourceId = request.getResourceId();
+
+        if (Strings.isNullOrEmpty(resourceId)) {
+            fetchAllResources(listener, client);
+        } else {
+            verifyAndFetchSingle(resourceId, listener, client);
+        }
+    }
+
+    private void fetchAllResources(ActionListener<GetResourceResponse> listener, ResourceSharingClient client) {
+        if (client == null) {
+            fetchResourcesByIds(null, listener);
+            return;
+        }
+
+        client.getAccessibleResourceIds(RESOURCE_INDEX_NAME, ActionListener.wrap(ids -> {
+            if (ids.isEmpty()) {
+                listener.onResponse(new GetResourceResponse(Collections.emptySet()));
+            } else {
+                fetchResourcesByIds(ids, listener);
```

Comment:
For real cases in sharable resource plugins, I think we should have another way of signaling to the plugin that the current user has access to everything instead of returning all ids. For users that can only see a subset then we should return what they can see. (The alternative to that would be the DLS approach, but that has been ruled out to make this forward looking for the remote metadata sdk).

Reply:
There is no way of knowing that user has access to everything unless they are super-admins or are mapped to all-access. Even today, the permissions are computed from mapped roles and evaluated against requested actions.

In this case sample plugin, security plugin will return all IDs in current index that are accessible. This can be expanded in future to evaluate associated permissions before fetching individual ids, however, that make its pretty open and invites room for mis-configuration. IMO, only admin certificates should have access to all resources.
To your suggestion, yes we can introduce a method like `hasAllAccess` that plugin's can leverage , in later iteration.